trantor/net/inner/BotanTLSConnectionImpl.h

```diff
@@ -0,0 +1,316 @@
+#pragma once
+
+#pragma once
+
+#include <botan/tls_server.h>
+#include <botan/tls_client.h>
+#include <botan/tls_callbacks.h>
+#include <botan/tls_policy.h>
+#include <botan/auto_rng.h>
+#include <botan/certstor.h>
+#include <botan/certstor_system.h>
+#include <botan/data_src.h>
+#include <botan/pkcs8.h>
+
+#include <trantor/net/TcpConnection.h>
+#include <trantor/utils/Logger.h>
+
+namespace trantor
+{
+class ClientCredentials : public Botan::Credentials_Manager
+{
+  public:
+    std::vector<Botan::Certificate_Store *> trusted_certificate_authorities(
+        const std::string &type,
+        const std::string &context) override
+    {
+        // return a list of certificates of CAs we trust for tls server
+        // certificates ownership of the pointers remains with
+        // Credentials_Manager
+        return {&certStore_};
+    }
+
+    std::vector<Botan::X509_Certificate> cert_chain(
+        const std::vector<std::string> &cert_key_types,
+        const std::string &type,
+        const std::string &context) override
+    {
+        // when using tls client authentication (optional), return
+        // a certificate chain being sent to the tls server,
+        // else an empty list
+        return {};
+    }
+
+    Botan::Private_Key *private_key_for(const Botan::X509_Certificate &cert,
+                                        const std::string &type,
+                                        const std::string &context) override
+    {
+        // when returning a chain in cert_chain(), return the private key
+        // associated with the leaf certificate here
+        return nullptr;
+    }
+    Botan::System_Certificate_Store certStore_;
+};
+
+class ServerCredentials : public Botan::Credentials_Manager
+{
+  public:
+    ServerCredentials(Botan::Private_Key *key, Botan::X509_Certificate *cert)
+        : key_(key), cert_(cert)
+    {
+    }
+
+    std::vector<Botan::Certificate_Store *> trusted_certificate_authorities(
+        const std::string &type,
+        const std::string &context) override
+    {
+        // if client authentication is required, this function
+        // shall return a list of certificates of CAs we trust
+        // for tls client certificates, otherwise return an empty list
+        return {};
+    }
+
+    std::vector<Botan::X509_Certificate> cert_chain(
+        const std::vector<std::string> &cert_key_types,
+        const std::string &type,
+        const std::string &context) override
+    {
+        // return the certificate chain being sent to the tls client
+        // e.g., the certificate file "botan.randombit.net.crt"
+        return {*cert_};
```

Comment:
You must return your server certificate only if the TLS server implementation specifically asked for a certificate whose subject public key has the type it requested. Otherwise, the certificate gets mis-associated internally and the ciphersuite selection will make mistakes.

Admittedly that's a footgun in the API.

```suggestion
        for(const auto& t : cert_key_types)
        {
            if(t == cert_->subject_public_key_algo().get_oid().to_formatted_string())
            {
                return {*cert_};
            }
        }

        return {};
```

Reply:
Thank you! This would have took me forever to figure out.
IMO updating the [example code](https://botan.randombit.net/handbook/api_ref/tls.html#code-example) would avoid and steer most people from hitting this landmine.